src/directives/when.ts

```diff
@@ -12,91 +12,63 @@
  *
- * html`
- *   when(checked, () => html`Checkmark is checked`, () => html`Checkmark is not
- * checked`);
- * `
- * ```
- *
- * @param condition the condition to test truthiness against
- * @param trueValue the value to render given a true condition
- * @param falseValue the value to render given a false condition
+ * @param condition the condition to check for truthiness
+ * @param caseMap object where keys are cases and values are functions which
+ *     return the value to render
+ * @param trueValue function that returns the value to render in case of
+ *     truthiness
+ * @param falseValue function that returns the value to render in case of
+ *     falsiness
  */
-export const when =
-    (condition: any, trueValue: () => any, falseValue: () => any):
-        Directive<NodePart> => directive((parentPart: NodePart) => {
-          let cache = partCaches.get(parentPart);
-
-          // Create a new cache if this is the first render
-          if (cache === undefined) {
-            // Cache consists of two parts, one for each condition, and a
-            // docment fragment which we cache the nodes of the condition that's
-            // not currently rendered.
-            cache = {
-              truePart: new NodePart(parentPart.options),
-              falsePart: new NodePart(parentPart.options),
-              cacheContainer: document.createDocumentFragment(),
-            };
-            partCaches.set(parentPart, cache);
-
-            cache.truePart.appendIntoPart(parentPart);
-            cache.falsePart.appendIntoPart(parentPart);
-          }
-
-          // Based on the condition, select which part to render and which value
-          // to set on that part.
-          const nextPart = condition ? cache.truePart : cache.falsePart;
-          const nextValue = condition ? trueValue() : falseValue();
-
-          // If we switched condition, swap nodes to/from the cache.
-          if (!!condition !== cache.prevCondition) {
-            // Get the part which was rendered for the opposite condition. This
-            // should be added to the cache.
-            const prevPart = condition ? cache.falsePart : cache.truePart;
+export function when(
+    condition: any, trueValue: WhenValue, falseValue?: WhenValue):
+    Directive<NodePart>;
+export function when(condition: any, caseMap: CaseMap): Directive<NodePart>;
+export function when(
+    condition: any,
+    trueValueOrCaseMap: WhenValue|CaseMap,
+    falseValue?: WhenValue): Directive<NodePart> {
+  let caseMap: CaseMap;
+  let trueValue: WhenValue;
 
-            // If the next part was rendered, take it from the cache
-            if (nextPart.value) {
-              parentPart.startNode.parentNode!.appendChild(
-                  cache.cacheContainer);
-            }
+  // test whether we are in case or in if/else mode
+  if (typeof trueValueOrCaseMap === 'object') {
+    caseMap = trueValueOrCaseMap;
+  } else {
+    trueValue = trueValueOrCaseMap;
+  }
```

Comment:
It is possible to reduce bytesize and improve performance here, if you are willing to use trueValueOrCaseMap for both cases.

```js
const caseMode = typeof trueValueOrCaseMap === 'object';

if  (caseMode) {
  nextValue = trueValueOrCaseMap[condition] || trueValueOrCaseMap.default;
// etc..
```

It makes the code slightly less readable, so it would require some code comments to clarify things. Maybe it would look better if the parameter has a different name? It's just a suggestion, I'm not sure if it's worth it.

Reply:
I had it like that initially, but it's much harder to read that way.